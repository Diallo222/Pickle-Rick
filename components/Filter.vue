<template lang="pug">
    .w-25
      h5 Find Characters
      //- form
        .mb-3
          label.form-label(for="name") Name
          input.form-control#name(v-model="name" placeholder="Enter character name")
        .mb-3
          label.form-label(for="status") Status
          select.form-select#status(:value="status")
            option(value="") -- Select Status --
            option(value="alive") Alive
            option(value="dead") Dead
            option(value="unknown") Unknown
        .mb-3
          label.form-label(for="species") Species
          select.form-select#species(v-model="species")
            option(value="") -- Select Specy --
            option(value="Human") Human
            option(value="Alien") Alien
            option(value="Robot") Robot
            option(value="Humanoid") Humanoid
            option(value="Cronenberg") Cronenberg
        .mb-3
          label.form-label(for="type") Type
          input.form-control#type(v-model="type" placeholder="Enter type")
        .mb-3
          label.form-label(for="gender") Gender
          select.form-select#gender(v-model="gender")
            option(value="") -- Select Gender --
            option(value="female") Female
            option(value="male") Male
            option(value="genderless") Genderless
            option(value="unknown") Unknown
      MultiSelector(:placeholder="'name'" :value="name" :options="selectorOptions" :onchange="updateFilter")
      h5.mt-2 Based on Location
      form(v-if="locations.length > 0 && !loading")
        .mb-3
          label.form-label(for="location") Location
          select.form-select#location(v-model="location")
            option(value="") -- Select Location --
            option(v-for="location in locations" :value="location.id") {{ location.name }}
      h5 Based on Episodes
        form(v-if="episodes.length > 0 && !loading")
          .mb-3
            label.form-label(for="episode") Episode
            select.form-select#episode(v-model="episode")
              option(value="") -- Select Episode --
              option(v-for="episode in episodes" :value="episode.id") {{ episode.name }}
      
</template>

<script setup lang="ts">
import { ref, watch } from "vue";
import { useCharactersStore } from "@/store/characters";
import { useLocationStore } from "~/store/locations";
import { useEpisodeStore } from "~/store/episode";

const props = defineProps({
  locations: Array,
  episodes: Array,
});

const name = ref("");
const status = ref("");
const species = ref("");
const type = ref("");
const gender = ref("");
const location = ref(0);
const episode = ref(0);

const charactersStore = useCharactersStore();
const locationStore = useLocationStore();
const episodeStore = useEpisodeStore();

const loading = computed(() => locationStore.isLoading);

const selectorOptions = [
  {
    label: "status",
    values: [" ","alive", "dead", "unknown"],
  },
  {
    label: "species",
    values: [" ","Human", "Alien", "Robot", "Humanoid", "Cronenberg"],
  },
  {
    label: "gender",
    values: [" ","female", "male", "genderless", "unknown"],
  },
];
// Apply filter
const applyFilters = () => {
  charactersStore.filterCharacters(
    name.value,
    status.value,
    species.value,
    type.value,
    gender.value
  );
};

const updateFilter = (key: string, value: string) => {
  switch (key) {
    case "name":
      name.value = value;
      break;
    case "type":
      type.value = value;
      break;
    case "status":
      status.value = value;
      break;
    case "species":
      species.value = value;
      break;
    case "gender":
      gender.value = value;
      break;
    default:
      break;
  }
  console.log("Filter updated:", key, value);

  applyFilters();
};

const findLocation = () => {
  if (!location.value) return (locationStore.location = {});
  locationStore.fetchLocation(location.value);
};

const findEpisode = () => {
  if (!episode.value) return (episodeStore.episode = {});
  episodeStore.fetchEpisode(episode.value);
};

// Watch for changes in filter values
watch(
  [name, status, species, type, gender, location, episode],
  ([name, status, species, type, gender, location, episode]) => {
    applyFilters();
  }
);

watch([location], ([location]) => findLocation());

watch([episode], ([episode]) => findEpisode());
</script>

<style></style>
