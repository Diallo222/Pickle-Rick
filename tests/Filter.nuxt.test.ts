import { describe, it, expect, vi, beforeEach } from "vitest";
import { shallowMount } from "@vue/test-utils";
import Filter from "@/components/Filter.vue";

const mockLocations = [
  { id: 1, name: "Earth" },
  { id: 2, name: "Earth (C-137)" },
  { id: 3, name: "Earth (Replacement Dimension)" },
  { id: 4, name: "unknown" },
];

describe("Filter Component", () => {
  let wrapper: any;
  let applyFiltersMock: any;
  let findLocationMock: any;

  beforeEach(() => {
    applyFiltersMock = vi.fn();
    findLocationMock = vi.fn();

    wrapper = shallowMount(Filter, {
      props: {
        locations: mockLocations,
      },
      methods: {
        applyFilters: applyFiltersMock,
        findLocation: findLocationMock,
      },
    });
  });

  it("renders correctly with given props", () => {
    expect(wrapper.element).toMatchSnapshot();
    const locationOptions = wrapper.findAll("#location option");
    expect(locationOptions.length).toBe(mockLocations.length + 1);
  });

  it("does not call findLocation when no location is selected", async () => {
    const locationSelect = wrapper.find("#location");
    await locationSelect.setValue("");
    expect(findLocationMock).not.toHaveBeenCalled();
  });
});
