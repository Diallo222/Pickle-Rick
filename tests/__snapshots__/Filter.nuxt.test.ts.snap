// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter Component > renders correctly 1`] = `
<div
  class="w-25"
>
  <h5>
    Find Characters
  </h5>
  bound HTMLFormElement {
    "0": <input
      class="form-control"
      id="name"
      placeholder="Enter character name"
    />,
    "1": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Status --
      </option>,
      "1": <option
        value="alive"
      >
        Alive
      </option>,
      "2": <option
        value="dead"
      >
        Dead
      </option>,
      "3": <option
        value="unknown"
      >
        Unknown
      </option>,
    },
    "2": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Specy --
      </option>,
      "1": <option
        value="Human"
      >
        Human
      </option>,
      "2": <option
        value="Alien"
      >
        Alien
      </option>,
      "3": <option
        value="Robot"
      >
        Robot
      </option>,
      "4": <option
        value="Humanoid"
      >
        Humanoid
      </option>,
      "5": <option
        value="Cronenberg"
      >
        Cronenberg
      </option>,
    },
    "3": <input
      class="form-control"
      id="type"
      placeholder="Enter type"
    />,
    "4": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Gender --
      </option>,
      "1": <option
        value="female"
      >
        Female
      </option>,
      "2": <option
        value="male"
      >
        Male
      </option>,
      "3": <option
        value="genderless"
      >
        Genderless
      </option>,
      "4": <option
        value="unknown"
      >
        Unknown
      </option>,
    },
  }
  <h5>
    Based on Location
  </h5>
  bound HTMLFormElement {
    "0": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Location --
      </option>,
      "1": <option
        value="1"
      >
        Earth
      </option>,
      "2": <option
        value="2"
      >
        Earth (C-137)
      </option>,
      "3": <option
        value="3"
      >
        Earth (Replacement Dimension)
      </option>,
      "4": <option
        value="4"
      >
        unknown
      </option>,
    },
  }
</div>
`;

exports[`Filter Component > renders correctly with given props 1`] = `
<div
  class="w-25"
>
  <h5>
    Find Characters
  </h5>
  bound HTMLFormElement {
    "0": <input
      class="form-control"
      id="name"
      placeholder="Enter character name"
    />,
    "1": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Status --
      </option>,
      "1": <option
        value="alive"
      >
        Alive
      </option>,
      "2": <option
        value="dead"
      >
        Dead
      </option>,
      "3": <option
        value="unknown"
      >
        Unknown
      </option>,
    },
    "2": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Specy --
      </option>,
      "1": <option
        value="Human"
      >
        Human
      </option>,
      "2": <option
        value="Alien"
      >
        Alien
      </option>,
      "3": <option
        value="Robot"
      >
        Robot
      </option>,
      "4": <option
        value="Humanoid"
      >
        Humanoid
      </option>,
      "5": <option
        value="Cronenberg"
      >
        Cronenberg
      </option>,
    },
    "3": <input
      class="form-control"
      id="type"
      placeholder="Enter type"
    />,
    "4": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Gender --
      </option>,
      "1": <option
        value="female"
      >
        Female
      </option>,
      "2": <option
        value="male"
      >
        Male
      </option>,
      "3": <option
        value="genderless"
      >
        Genderless
      </option>,
      "4": <option
        value="unknown"
      >
        Unknown
      </option>,
    },
  }
  <h5>
    Based on Location
  </h5>
  bound HTMLFormElement {
    "0": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Location --
      </option>,
      "1": <option
        value="1"
      >
        Earth
      </option>,
      "2": <option
        value="2"
      >
        Earth (C-137)
      </option>,
      "3": <option
        value="3"
      >
        Earth (Replacement Dimension)
      </option>,
      "4": <option
        value="4"
      >
        unknown
      </option>,
    },
  }
</div>
`;

exports[`Filter component > renders correctly 1`] = `
<div
  class="w-25"
>
  <h5>
    Find Characters
  </h5>
  bound HTMLFormElement {
    "0": <input
      class="form-control"
      id="name"
      placeholder="Enter character name"
    />,
    "1": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Status --
      </option>,
      "1": <option
        value="alive"
      >
        Alive
      </option>,
      "2": <option
        value="dead"
      >
        Dead
      </option>,
      "3": <option
        value="unknown"
      >
        Unknown
      </option>,
    },
    "2": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Specy --
      </option>,
      "1": <option
        value="Human"
      >
        Human
      </option>,
      "2": <option
        value="Alien"
      >
        Alien
      </option>,
      "3": <option
        value="Robot"
      >
        Robot
      </option>,
      "4": <option
        value="Humanoid"
      >
        Humanoid
      </option>,
      "5": <option
        value="Cronenberg"
      >
        Cronenberg
      </option>,
    },
    "3": <input
      class="form-control"
      id="type"
      placeholder="Enter type"
    />,
    "4": bound HTMLSelectElement {
      "0": <option
        value=""
      >
        -- Select Gender --
      </option>,
      "1": <option
        value="female"
      >
        Female
      </option>,
      "2": <option
        value="male"
      >
        Male
      </option>,
      "3": <option
        value="genderless"
      >
        Genderless
      </option>,
      "4": <option
        value="unknown"
      >
        Unknown
      </option>,
    },
  }
  <h5>
    Based on Location
  </h5>
  <!--v-if-->
</div>
`;
